components/raftstore/src/store/peer.rs

```diff
@@ -3471,6 +3514,118 @@ fn make_transfer_leader_response() -> RaftCmdResponse {
     resp
 }
 
+const REQUEST_FLAG: u8 = b'r';
+const LOCKED_FLAG: u8 = b'l';
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct ReadIndexContext {
+    pub id: Uuid,
+    pub request: Option<raft_cmdpb::ReadIndexRequest>,
+    pub locked: Option<LockInfo>,
+}
+
+impl ReadIndexContext {
+    pub fn parse(bytes: &[u8]) -> Result<ReadIndexContext> {
+        use tikv_util::codec::number::*;
+
+        if bytes.len() < 16 {
+            return Err(box_err!("read index context less than 16 bytes"));
+        }
+        let mut res = ReadIndexContext {
+            id: Uuid::from_slice(&bytes[..16]).unwrap(),
+            request: None,
+            locked: None,
+        };
+        let mut bytes = &bytes[16..];
+        while !bytes.is_empty() {
+            match read_u8(&mut bytes).unwrap() {
+                REQUEST_FLAG => {
+                    let len = decode_var_u64(&mut bytes)? as usize;
+                    let mut request = raft_cmdpb::ReadIndexRequest::default();
+                    request.merge_from_bytes(&bytes[..len])?;
+                    bytes = &bytes[len..];
+                    res.request = Some(request);
+                }
+                LOCKED_FLAG => {
+                    let len = decode_var_u64(&mut bytes)? as usize;
+                    let mut locked = LockInfo::default();
+                    locked.merge_from_bytes(&bytes[..len])?;
+                    bytes = &bytes[len..];
+                    res.locked = Some(locked);
+                }
+                // just break for forward compatibility
+                _ => break,
+            }
+        }
+        Ok(res)
+    }
+
+    pub fn to_bytes(&self) -> Vec<u8> {
+        Self::fields_to_bytes(self.id, self.request.as_ref(), self.locked.as_ref())
+    }
+
+    pub fn fields_to_bytes(
+        id: Uuid,
+        request: Option<&raft_cmdpb::ReadIndexRequest>,
+        locked: Option<&LockInfo>,
+    ) -> Vec<u8> {
+        let request_size = request.map(Message::compute_size);
+        let locked_size = locked.map(Message::compute_size);
+        let field_size = |s: Option<u32>| s.map(|s| 1 + MAX_VAR_U64_LEN + s as usize).unwrap_or(0);
```

Comment:
```suggestion
        let field_size = |s: Option<u32>| s.map(|s| 8 + MAX_VAR_U64_LEN + s as usize).unwrap_or(0);
```
Looks like the flag is a `u8`

Reply:
`u8` is one byte.

Reply:
Yes, indeed